File: src/main/java/sprykerFileCreator/model/ModelFactory.java
package pav.sprykerFileCreator.model;

import com.intellij.openapi.project.Project;
import pav.sprykerFileCreator.config.SprykerPluginConfig;
import pav.sprykerFileCreator.model.command.UpdateDocBlockCommand;
import pav.sprykerFileCreator.model.definition.DefinitionProvider;
import pav.sprykerFileCreator.model.definition.DefinitionProviderInterface;
import pav.sprykerFileCreator.model.finder.ClassFinder;
import pav.sprykerFileCreator.model.finder.ClassFinderInterface;
import pav.sprykerFileCreator.model.generator.*;
import pav.sprykerFileCreator.model.manager.ClassManager;
import pav.sprykerFileCreator.model.manager.ClassManagerInterface;
import pav.sprykerFileCreator.model.matcher.ClassTypeMatcher;
import pav.sprykerFileCreator.model.matcher.ClassTypeMatcherInterface;
import pav.sprykerFileCreator.model.renderer.PhpClassRenderer;
import pav.sprykerFileCreator.model.renderer.PhpClassRendererInterface;
import pav.sprykerFileCreator.model.resolver.ClassResolver;
import pav.sprykerFileCreator.model.resolver.ClassResolverInterface;
import pav.sprykerFileCreator.model.writer.FileWriter;
import pav.sprykerFileCreator.model.writer.FileWriterInterface;

public class ModelFactory {

    private final DefinitionProviderInterface definitionProvider;
    private ClassResolverInterface classResolver;

    public ModelFactory() {
        this.definitionProvider = this.createDefinitionProvider();
    }

    public ClassManagerInterface createClassManager(Project project, String projectName) {
        return new ClassManager(
                this.createClassGenerator(project, projectName),
                this.createClassRenderer(),
                this.createFileWriter(project)
        );
    }

    public ClassTypeMatcherInterface createClassTypeMatcher() {
        return new ClassTypeMatcher(this.definitionProvider);
    }

    public UpdateDocBlockCommand createUpdateDocBlockCommand(Project project, String projectName) {
        return new UpdateDocBlockCommand(
                this.createClassTypeMatcher(),
                this.createDocBlockGenerator(project, projectName),
                this.createClassRenderer()
        );
    }

    public DefinitionProviderInterface createDefinitionProvider() {
        return new DefinitionProvider();
    }

    private ClassResolverInterface createClassResolver(Project project, String projectName) {
        if (this.classResolver == null) {
            this.classResolver = new ClassResolver(
                    this.getConfig(project, projectName).getProjectName(),
                    this.getConfig(project, projectName).getCoreNames(),
                    this.createClassFinder(project),
                    this.definitionProvider
            );
        }

        return this.classResolver;
    }

    private ClassFinderInterface createClassFinder(Project project) {
        return new ClassFinder(project);
    }

    private ClassGeneratorInterface createClassGenerator(Project project, String projectName) {
        return new ClassGenerator(
                this.createParentGenerator(project, projectName),
                this.createDocBlockGenerator(project, projectName),
                this.createDefinitionProvider()
        );
    }

    private PhpClassRendererInterface createClassRenderer() {
        return new PhpClassRenderer();
    }

    private FileWriterInterface createFileWriter(Project project) {
        return new FileWriter(project);
    }

    private ParentGeneratorInterface createParentGenerator(Project project, String projectName) {
        return new ParentGenerator(
                this.createClassResolver(project, projectName),
                this.createClassFinder(project)
        );
    }

    private DocBlockGeneratorInterface createDocBlockGenerator(Project project, String projectName) {
        return new DocBlockGenerator(
                this.createClassResolver(project, projectName),
                this.definitionProvider
        );
    }

    private SprykerPluginConfig getConfig(Project project, String projectName) {
        return SprykerPluginConfig.getInstance(project, projectName);
    }

}

File: src/main/kotlin/com/github/tobi812/sprykerplugin/models/ModelFactory.kt
package com.github.tobi812.sprykerplugin.models

import com.github.tobi812.sprykerplugin.config.SprykerPluginConfig
import com.github.tobi812.sprykerplugin.models.command.AddMissingFactoryMethodsCommand
import com.github.tobi812.sprykerplugin.models.command.UpdateDocBlockCommand
import com.github.tobi812.sprykerplugin.models.definitions.DefinitionProvider
import com.github.tobi812.sprykerplugin.models.definitions.DefinitionProviderInterface
import com.github.tobi812.sprykerplugin.models.finder.ClassFinder
import com.github.tobi812.sprykerplugin.models.finder.ClassFinderInterface
import com.github.tobi812.sprykerplugin.models.finder.FactoryFinder
import com.github.tobi812.sprykerplugin.models.finder.MethodFinderInterface
import com.github.tobi812.sprykerplugin.models.generator.*
import com.github.tobi812.sprykerplugin.models.manager.ClassManager
import com.github.tobi812.sprykerplugin.models.manager.ClassManagerInterface
import com.github.tobi812.sprykerplugin.models.matcher.ClassTypeMatcher
import com.github.tobi812.sprykerplugin.models.matcher.ClassTypeMatcherInterface
import com.github.tobi812.sprykerplugin.models.renderer.PhpClassRenderer
import com.github.tobi812.sprykerplugin.models.renderer.PhpClassRendererInterface
import com.github.tobi812.sprykerplugin.models.resolver.ClassResolver
import com.github.tobi812.sprykerplugin.models.resolver.ClassResolverInterface
import com.github.tobi812.sprykerplugin.models.writer.FileWriter
import com.github.tobi812.sprykerplugin.models.writer.FileWriterInterface
import com.intellij.openapi.project.Project

class ModelFactory {
    val definitionProvider: DefinitionProviderInterface
    val classTypeMatcher: ClassTypeMatcherInterface

    init {
        this.definitionProvider = DefinitionProvider()
        this.classTypeMatcher = ClassTypeMatcher(this.definitionProvider)
    }

    fun createClassManager(project: Project, projectName: String): ClassManagerInterface {
        return ClassManager(
                this.createClassGenerator(project, projectName),
                this.createClassRenderer(),
                this.createFileWriter(project)
        )
    }

    fun createUpdateDocBlockCommand(project: Project, projectName: String): UpdateDocBlockCommand {
        return UpdateDocBlockCommand(
                this.classTypeMatcher,
                this.createDocBlockGenerator(project, projectName),
                this.createClassRenderer()
        )
    }

    fun createAddMissingFactoryMethodsCommand(): AddMissingFactoryMethodsCommand {
        return AddMissingFactoryMethodsCommand(this.createClassRenderer(), this.classTypeMatcher)
    }

    fun createFileWriter(project: Project): FileWriterInterface {
        return FileWriter(project)
    }

    fun createClassFinder(project: Project): ClassFinderInterface {
        return ClassFinder(project)
    }

    fun createMethodFinder(project: Project, projectName: String): MethodFinderInterface {
        return FactoryFinder(
            this.createClassGenerator(project, projectName),
            this.createClassFinder(project)
        )
    }

    private fun createClassGenerator(project: Project, projectName: String): ClassGeneratorInterface {
        return ClassGenerator(
            this.createParentGenerator(project, projectName),
            this.createDocBlockGenerator(project, projectName),
            this.definitionProvider
        )
    }

    private fun createClassResolver(project: Project, projectName: String): ClassResolverInterface {
        val config: SprykerPluginConfig = SprykerPluginConfig.getInstance(project)

        return ClassResolver(
            config.projectName,
            config.coreNames,
            this.createClassFinder(project),
            this.definitionProvider
        )
    }

    private fun createClassRenderer(): PhpClassRendererInterface {
        return PhpClassRenderer()
    }

    private fun createParentGenerator(project: Project, projectName: String): ParentGeneratorInterface {
        return ParentGenerator(
                this.createClassResolver(project, projectName),
                this.createClassFinder(project)
        )
    }

    private fun createDocBlockGenerator(project: Project, projectName: String): DocBlockGeneratorInterface {
        return DocBlockGenerator(
                this.createClassResolver(project, projectName),
                this.definitionProvider
        )
    }
}
